features/login/impl/src/main/kotlin/io/element/android/features/login/impl/di/QrCodeLoginComponent.kt

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2024 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.element.android.features.login.impl.di
+
+import com.squareup.anvil.annotations.ContributesTo
+import com.squareup.anvil.annotations.MergeSubcomponent
+import dagger.Subcomponent
+import io.element.android.libraries.architecture.NodeFactoriesBindings
+import io.element.android.libraries.di.AppScope
+import io.element.android.libraries.di.SingleIn
+
+@SingleIn(QrCodeLoginScope::class)
+@MergeSubcomponent(QrCodeLoginScope::class)
+interface QrCodeLoginComponent : NodeFactoriesBindings {
```

Comment:
Do we really need a separate component and scope for this? This is just to keep the `StateFlow<QrCodeLoginStep>`?

Reply:
The problem here is `QrCodeLoginManager` had to be shared by several components (`QrCodeLofingFlowNode`, `QrCodeScanPresenter`) and for that I needed a singleton if I'm not mistaken. But having a singleton can be troublesome if you aren't super careful with cleaning up its state once you're done working with it, so I created a new component to hold it that will be created when the QR code flow starts and discarded once it's finished.

This should ensure the object is shared but can't be leaked in any way, although it's more complex than I intended it to be.